### PR TITLE
Use DictionaryArray's iterator in `compare_dict_op`

### LINKED ISSUE #1329: Use DictionaryArray's iterator in `compare_dict_op`
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
(This section helps Arrow developers understand the context and *why* for this feature, in addition to  the *what*)

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by viirya**
# Which issue does this PR close?

Closes #1329.

# Rationale for this change
 
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: arrow/src/compute/kernels/comparison.rs [hunk 1]
```diff
@@ -2213,27 +2213,26 @@ macro_rules! compare_dict_op {
                     .to_string(),
             ));
         }
-        let left_values = $left.values().as_any().downcast_ref::<$value_ty>().unwrap();
-        let right_values = $right
+
+        let left_iter = $left
```
Comment (alamb) [inline arrow/src/compute/kernels/comparison.rs:2217]: This sure looks much nicer 👍

**Review by alamb — APPROVED** [review]
The only potential concern I have here is that the `take_iter()s` use `value()` instead of `value_unchecked()` -- though perhaps we can merge this PR in and then add some additional unsafe / skipping bounds checking if it turns out to be in anyone's critical path / shows up in profiles

Thanks @viirya

**Comment by alamb** [comment]
> Thanks @alamb . As we don't know if the keys in key iterator are fully in the valid range,seems we cannot skipping bounds check in take_iter? 

That is true in general. However, if we are iterating over a set of keys from a known valid `DictionaryArray` the indexes are known to be valid

> If it is critical, maybe we can have take_iter_unchecked in the future if it is necessary?

I think this is wise strategy

**Comment by alamb** [comment]
I filed https://github.com/apache/arrow-rs/issues/1371 for follow on performance work and I am merging this one in now 🚀  

thanks @viirya

**Comment by viirya** [comment]
Thank you @alamb !

**Pull request merged**